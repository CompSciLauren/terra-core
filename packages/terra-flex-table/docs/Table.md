# Terra FlexTable

The Terra FlexTable is a structural component used to create standard html tables.

## Getting Started

- Install with [npmjs](https://www.npmjs.com):
  - `npm install terra-flex-table`
  - `yarn add terra-flex-table`

## Usage

```jsx
import React from 'react';
import Table, { Row, Cell } from 'terra-flex-table';

<Table>
  <Row isSelectable key="row-0">
    <Cell key="row-0-cell-0">
      Table row text
    </Cell>
    <Cell key="row-0-cell-0">
      Table row text
    </Cell>
  </Row>
  <Row isSelectable key="row-1">
    <Cell key="row-1-cell-0">
      Table row text
    </Cell>
    <Cell key="row-1-cell-1">
      Table row text
    </Cell>
  </Row>
</Table>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
