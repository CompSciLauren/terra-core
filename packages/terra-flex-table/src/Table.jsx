import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import ResizeObserver from 'resize-observer-polyfill';
import styles from './Table.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The children passed to the component
   */
  children: PropTypes.node,
  /**
   * The header passed to the table
   */
  headerCells: PropTypes.arrayOf(PropTypes.element),
  /**
   * The padding styling to apply to the child list item content.
   * One of `'none'`, `'standard'`, `'compact'`.
   */
  paddingStyle: PropTypes.oneOf(['none', 'standard', 'compact']),
  /**
   * Function callback for the ref of the table.
   */
  refCallback: PropTypes.func,
  /**
   * Whether or not the table should expanded to fill its parent element.
   */
  fill: PropTypes.bool,
};

const defaultProps = {
  headerCells: [],
  paddingStyle: 'none',
  fill: false,
};

class Table extends React.Component {
  constructor(props) {
    super(props);

    this.updateSize = this.updateSize.bind(this);
    this.initializeResize = this.initializeResize.bind(this);
    this.removeResize = this.removeResize.bind(this);
    this.bodyRef = React.createRef();
    this.headerRef = React.createRef();
    this.animationFrameID = null;
    this.offset = 0;
  }

  componentDidMount() {
    if (this.props.fill && this.bodyRef.current) {
      this.initializeResize();
    }
  }

  componentDidUpdate() {
    if (this.props.fill && this.bodyRef.current) {
      this.initializeResize();
    } else {
      this.removeResize();
    }
  }

  componentWillUnmount() {
    if (this.props.fill && this.bodyRef.current) {
      this.removeResize();
    }
  }

  initializeResize() {
    if (!this.resizeListenerAdded) {
      this.resizeObserver = new ResizeObserver(() => {
        const offset = this.bodyRef.current.offsetWidth - this.bodyRef.current.clientWidth;
        if (this.offset !== offset) {
          this.offset = offset;
          this.animationFrameID = window.requestAnimationFrame(() => {
            this.animationFrameID = null;
            this.updateSize(this.offset);
          });
        }
      });
      this.resizeObserver.observe(this.bodyRef.current);
      this.resizeListenerAdded = true;
    }
  }

  removeResize() {
    if (this.resizeListenerAdded) {
      window.cancelAnimationFrame(this.animationFrameID);
      this.resizeObserver.disconnect(this.bodyRef.current);
      this.resizeListenerAdded = false;
    }
  }

  updateSize(width) {
    if (this.headerRef.current) {
      this.headerRef.current.style.paddingRight = `${width}px`;
    }
  }

  render() {
    const {
      children,
      fill,
      headerCells,
      paddingStyle,
      refCallback,
      ...customProps
    } = this.props;
    const attrSpread = {};
    if (paddingStyle !== 'none') {
      attrSpread['data-table-padding'] = paddingStyle;
    }

    let header;
    let makeInline = false;
    if (headerCells.length) {
      makeInline = headerCells.every(cell => (cell.props.width && cell.props.width.static));
      header = (
        <div className={cx(['header'])} role="rowgroup" ref={this.headerRef}>
          <div className={cx(['header-content'])} role="row">
            {headerCells}
          </div>
        </div>
      );
    }

    const tableClassNames = cx([
      'table',
      { fill },
      { 'is-inline': makeInline },
      customProps.className,
    ]);

    return (
      <div {...customProps} {...attrSpread} className={tableClassNames} ref={refCallback} role="grid">
        {header}
        <div className={cx(['body'])} role="rowgroup" ref={this.bodyRef}>
          {children}
        </div>
      </div>
    );
  }
}

Table.propTypes = propTypes;
Table.defaultProps = defaultProps;

export default Table;
