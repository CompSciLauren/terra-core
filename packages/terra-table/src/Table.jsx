import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import ContentContainer from 'terra-content-container';
import styles from './Table.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The children passed to the component
   */
  children: PropTypes.node,
  /**
   * The HeaderRow passed to the table, should have widths aligned with the child Rows.
   */
  headerRow: PropTypes.element,
  /**
   * Element to append to the top of the table row. i.e. toolbars etc.
   */
  headerNode: PropTypes.node,
  /**
   * Whether or not the table should expanded to fill its parent element.
   */
  fill: PropTypes.bool,
  /**
   * Element to append to the bottom of the table row. i.e. toolbars etc.
   */
  footerNode: PropTypes.node,
  /**
   * The padding styling to apply to the child list item content.
   * One of `'none'`, `'standard'`, `'compact'`.
   */
  paddingStyle: PropTypes.oneOf(['none', 'standard', 'compact']),
  /**
   * Function callback for the ref of the table.
   */
  scrollRefCallback: PropTypes.func,
};

const defaultProps = {
  paddingStyle: 'none',
  fill: false,
};

const Table = ({
  children,
  fill,
  footerNode,
  headerRow,
  headerNode,
  paddingStyle,
  scrollRefCallback,
  ...customProps
}) => {
  const attrSpread = {};
  if (paddingStyle !== 'none') {
    attrSpread['data-table-padding'] = paddingStyle;
  }

  const makeInline = headerRow ? React.Children.toArray(headerRow.props.children).every(cell => (cell.props.width && cell.props.width.static)) : undefined;
  const hasEndNodes = headerNode || footerNode;

  const tableClassNames = cx([
    'table',
    { fill },
    { 'is-inline': makeInline },
    { rounded: !hasEndNodes },
    customProps.className,
  ]);

  const rows = (
    <div {...customProps} {...attrSpread} className={tableClassNames} role="grid">
      {headerRow}
      {children ? (
        <div className={cx(['body'])} role="rowgroup" ref={scrollRefCallback}>
          {children}
        </div>
      ) : undefined}
    </div>
  );

  if (!hasEndNodes) {
    return rows;
  }

  return (
    <ContentContainer
      fill={fill}
      footer={footerNode}
      header={headerNode}
      className={cx(
        'rounded',
      )}
    >
      {rows}
    </ContentContainer>
  );
};

Table.propTypes = propTypes;
Table.defaultProps = defaultProps;

export default Table;
