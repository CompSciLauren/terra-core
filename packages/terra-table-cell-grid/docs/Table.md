# Terra TableCellGrid

The Terra TableCellGrid is a structural component used to create standard html tables.

## Getting Started

- Install with [npmjs](https://www.npmjs.com):
  - `npm install terra-table-cell-grid`
  - `yarn add terra-table-cell-grid`

## Usage

```jsx
import React from 'react';
import Table, { Row, Cell } from 'terra-table-cell-grid';

<TableCellGrid>
  <Row isSelectable key="row-0">
    <Cell key="row-0-cell-0">
      Table row text
    </Cell>
    <Cell key="row-0-cell-0">
      Table row text
    </Cell>
  </Row>
  <Row isSelectable key="row-1">
    <Cell key="row-1-cell-0">
      Table row text
    </Cell>
    <Cell key="row-1-cell-1">
      Table row text
    </Cell>
  </Row>
</TableCellGrid>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
