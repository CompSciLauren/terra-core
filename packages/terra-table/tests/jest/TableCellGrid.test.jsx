import React from 'react';
import Table from '../../src/Table';

// Snapshot test
it('should render a default table', () => {
  const shallowComponent = shallow(
    <Table />,
  );
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a table with standard padding', () => {
  const shallowComponent = shallow(
    <Table
      paddingStyle="standard"
    />,
  );
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a table with compact padding', () => {
  const shallowComponent = shallow(
    <Table
      paddingStyle="compact"
    />,
  );
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a table with fill style', () => {
  const shallowComponent = shallow(
    <Table
      fill
    />,
  );
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a table with header/footer nodes', () => {
  const shallowComponent = shallow(
    <Table
      headerNode={<p>header</p>}
      footerNode={<p>footer</p>}
    />,
  );
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a table with refCallback', () => {
  const shallowComponent = shallow(
    <Table
      refCallback={jest.fn()}
    />,
  );
  expect(shallowComponent).toMatchSnapshot();
});
