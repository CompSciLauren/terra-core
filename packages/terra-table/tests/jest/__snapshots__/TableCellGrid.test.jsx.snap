// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a default table cell grid 1`] = `
<div
  className="table rounded"
  role="grid"
/>
`;

exports[`should render a table cell grid with compact padding 1`] = `
<div
  className="table rounded"
  data-table-padding="compact"
  role="grid"
/>
`;

exports[`should render a table cell grid with fill style 1`] = `
<div
  className="table fill rounded"
  role="grid"
/>
`;

exports[`should render a table cell grid with header/footer nodes 1`] = `
<ContentContainer
  className="rounded"
  fill={false}
  footer={
    <p>
      footer
    </p>
  }
  header={
    <p>
      header
    </p>
  }
>
  <div
    className="table"
    role="grid"
  />
</ContentContainer>
`;

exports[`should render a table cell grid with refCallback 1`] = `
<div
  className="table rounded"
  refCallback={[MockFunction]}
  role="grid"
/>
`;

exports[`should render a table cell grid with standard padding 1`] = `
<div
  className="table rounded"
  data-table-padding="standard"
  role="grid"
/>
`;
